code/modules/telesci/dses.dm

```diff
@@ -0,0 +1,291 @@
+		if(module_in_list("getDST"))
+			t += "<BR><B>Distance from previous ping: [last_distance]</B>"
+		if(module_in_list("crdLOG") && positive_locations.len)
+			t += "<BR>==Logged successful pings=="
+			var/current_loc_itt
+			for(var/atom/A in positive_locations)
+				current_loc_itt ++
+				t += "<BR>\the [A] in [A.loc] at [A.x-WORLD_X_OFFSET[A.z]] - [A.y-WORLD_Y_OFFSET[A.z]] - [A.z]"
+				t += " === <A href='?src=\ref[src];wipe=[current_loc_itt]'>Wipe log</A>"
+			t += "<BR><A href='?src=\ref[src];wipe=0'>Wipe All</A>"
+		t += "<BR><B>Modules installed</B>"
+		if(module_list.len)
+			var/current_module_num = 0
+			for(var/obj/item/weapon/dses_module/D in module_list)
+				current_module_num ++
+				t += "<BR>Module: [D.module_name]"
+				t += "<BR><A href='?src=\ref[src];eject=[current_module_num]'>Eject Module</A>"
+		else
+			t += "<BR>No modules detected"
+		t += "<BR>Module slots [module_list.len]/[module_limit]"
+		t += "<BR><A href='?src=\ref[src];eject_cell=1'>Eject Cell</A>"
+	else
+		t += "<B>No cell detected.</B>"
+	var/datum/browser/popup = new(user, "DSES", name, 300, 450)
+	popup.set_content(t)
+	popup.set_title_image(user.browse_rsc_icon(src.icon, src.icon_state))
+	popup.open()
+
+/obj/item/device/dses/Topic(href, href_list)
+	..()
+	usr.set_machine(src)
+	if(href_list["pulse"])
+		if (usr.get_active_hand() != src || usr.stat)
+			to_chat(usr, "<span class = 'caution'>You need to have \the [src] in your hand to do that!</span>")
+			return
+		if(C.charge < pulse_cost)
+			to_chat(usr, "<span class = 'caution'>\The [C] has insufficient charge to support another pulse.</span>")
+			return
+		handle_pulse(usr)
+
+	else if(href_list["eject"])
+		var/index = text2num(href_list["eject"])
+		if(index && index <= module_list.len)
+			var/obj/item/weapon/dses_module/D = module_list[index]
+			uninstall_module(D, usr)
+			D = null
+	else if(href_list["wipe"])
+		var/index = text2num(href_list["wipe"])
+		if(index && index <= positive_locations.len)
+			var/atom/A = positive_locations[index]
+			positive_locations -= A
+		else
+			positive_locations = list()
+	else if(href_list["tog_pulse"])
+		auto_pulse = !auto_pulse
+	else if(href_list["eject_cell"])
+		if(C)
+			C.forceMove(get_turf(src))
+			C = null
+			usr << browse(null, "window=DSES")
+			return
+	updateSelfDialog(usr)
+
+/obj/item/device/dses/updateSelfDialog(mob/user)
+	src.attack_self(user)
+
+/obj/item/device/dses/process()
+	if(auto_pulse && (!last_pulse || last_pulse <= world.time+pulse_between))
+		handle_pulse()
+
+/obj/item/device/dses/proc/handle_pulse(mob/user)
+	if(!last_pulse || last_pulse <= world.time+pulse_between)
+		var/atom/detected
+		detected = pulse()
+
+		if(!detected)
+			if(user)
+				to_chat(user, "<span class = 'warning'>No structures detected.</span>")
+			playsound(src, 'sound/machines/buzz-sigh.ogg', 50, 1, gas_modified = 0)
+			return
+
+		playsound(src, 'sound/machines/ping.ogg', 50, 1, gas_modified = 0)
+
+		if(module_in_list("dirGET"))
```

Comment:
Defines are preprocessor replacement systems - you can make a define for a string and the string gets subbed in at compile time. You use a define because the compiler treats them as identifiers and catches typoes, which strings do not do.

Use a define.